homeassistant/components/proximity/sensor.py

```diff
@@ -0,0 +1,129 @@
+"""Support for Proximity sensors."""
+
+from __future__ import annotations
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+)
+from homeassistant.const import CONF_NAME, UnitOfLength
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import ATTR_DIR_OF_TRAVEL, ATTR_DIST_TO, ATTR_NEAREST, DOMAIN
+from .coordinator import ProximityDataUpdateCoordinator
+
+SENSORS_PER_ENTITY: list[SensorEntityDescription] = [
+    SensorEntityDescription(
+        key=ATTR_DIST_TO,
+        device_class=SensorDeviceClass.DISTANCE,
+        native_unit_of_measurement=UnitOfLength.METERS,
+    ),
+    SensorEntityDescription(
+        key=ATTR_DIR_OF_TRAVEL,
+        translation_key=ATTR_DIR_OF_TRAVEL,
+        icon="mdi:compass-outline",
+        device_class=SensorDeviceClass.ENUM,
+        options=[
+            "arrived",
+            "away_from",
+            "stationary",
+            "towards",
+        ],
+    ),
+]
+
+SENSORS_PER_PROXIMITY: list[SensorEntityDescription] = [
+    SensorEntityDescription(
+        key=ATTR_NEAREST, translation_key=ATTR_NEAREST, icon="mdi:near-me"
+    ),
+]
+
+
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up the Proximity sensor platform."""
+    if discovery_info is None:
+        return
+
+    coordinator: ProximityDataUpdateCoordinator = hass.data[DOMAIN][
+        discovery_info[CONF_NAME]
+    ]
+
+    entities: list[ProximitySensor | ProximityTrackedEntitySensor] = [
+        ProximitySensor(description, coordinator)
+        for description in SENSORS_PER_PROXIMITY
+    ]
+
+    entities += [
+        ProximityTrackedEntitySensor(description, coordinator, tracked_entity_id)
+        for description in SENSORS_PER_ENTITY
+        for tracked_entity_id in coordinator.proximity_devices
+    ]
+
+    async_add_entities(entities)
+
+
+class ProximitySensor(SensorEntity, CoordinatorEntity[ProximityDataUpdateCoordinator]):
+    """Represents a Proximity sensor."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        description: SensorEntityDescription,
+        coordinator: ProximityDataUpdateCoordinator,
+    ) -> None:
+        """Initialize the proximity."""
+        super().__init__(coordinator)
+
+        self.entity_description = description
+
+        self._attr_name = f"{coordinator.friendly_name} {description.key}"
```

Comment:
I think we should set a description name instead of the key and use here, even if it's temporary until we have a config entry. Add a comment that we'll remove the entity name when we have a config entry and can follow the entity naming guidelines.

Reply:
7ffe810